src/main/java/com/thebois/models/world/generation/noises/PerlinNoise.java

```diff
@@ -0,0 +1,179 @@
+package com.thebois.models.world.generation.noises;
+
+/**
+ * Generator used to generate Perlin Noise.
+ */
+public class PerlinNoise implements INoise {
+
+    private int currentOctave;
+    private final int octaves;
+    private final float amplitude;
+    private final double frequency;
+    private final double persistence;
+    private int seed = 0;
+    private final int[] primeNumberArray = {
+        15731,
+        789221,
+        1376312589,
+        547063007,
+        522882781,
+        286040101,
+        628595861,
+        426993577,
+        360599501,
+        795799423,
+        185183639,
+        719087363,
+        249311663,
+        };
+
+    /**
+     * Instantiate a Perlin Noise Generator with given arguments.
+     *
+     * @param octaves     Number of octaves used for generating perlin noise, A number equal or
+     *                    greater than 1.
+     * @param amplitude   The amplitude used to amplify the resulting noise. The value should be not
+     *                    be 0.
+     * @param frequency   Used to decide how big the square should be.
+     * @param persistence Used to decide much the noise should change per addition. The value should
+     *                    not be 0 if octave is greater than 1;
+     */
+    public PerlinNoise(
+        final int octaves,
+        final float amplitude,
+        final double frequency,
+        final double persistence) {
+        this.octaves = octaves;
+        this.amplitude = amplitude;
+        this.frequency = frequency;
+        this.persistence = persistence;
+    }
+
+    /**
+     * Generates a float value with Perlin Noise algorithm given a position.
+     *
+     * @param coordinateX X coordinate for the position.
+     * @param coordinateY Y coordinate for the position.
+     *
+     * @return The Perlin Noise value.
+     */
+    public float sample(final float coordinateX, final float coordinateY) {
+        double total = 0;
+        for (int i = 0; i < octaves; i++) {
+            currentOctave = i;
+            final double octaveAmplification = Math.pow(persistence, i);
+            final double frequencyOffSet = frequency * Math.pow(2, i);
+
+            total = total + interpolateNoise(coordinateX * frequencyOffSet + seed,
+                                             coordinateY * frequencyOffSet + seed)
+                            * octaveAmplification;
+        }
+        return (float) (total * amplitude);
+    }
+
+    private double interpolateNoise(final double coordinateX, final double coordinateY) {
+        // Convert to integers
+        final int integerX = (int) coordinateX;
+        final int integerY = (int) coordinateY;
```

Comment:
Question

Is java smart enough to cast these for us if the methods takes ints instead of doubles?